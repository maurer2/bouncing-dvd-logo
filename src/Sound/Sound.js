import React, { useState, useEffect, useRef, useContext } from 'react';
import PropTypes from 'prop-types';
import soundFile from './soundFile.wav';

import Store from '../Store';

const Sound = ({ playSound }) => {
  const [soundIsPlaying, setSoundIsPlaying] = useState(false);
  const prevPlaySound = useRef(false);
  const { soundIsDisabled } = useContext(Store);

  useEffect(() => {
    if (soundIsDisabled) {
      return;
    }

    if (playSound && playSound !== prevPlaySound.current) {
      setSoundIsPlaying(true);

      window.setTimeout(() => {
        setSoundIsPlaying(false);
      }, 800);
    }

    prevPlaySound.current = prevPlaySound;
  }, [playSound, soundIsDisabled]);

  return (
    <>
      {soundIsPlaying
        ? (
          <audio autoPlay>
            <source src={ soundFile } type="audio/wav" />
          </audio>
        )
        : null
      }
    </>
  );
};

const { bool } = PropTypes;

Sound.propTypes = {
  playSound: bool,
};

export default Sound;
