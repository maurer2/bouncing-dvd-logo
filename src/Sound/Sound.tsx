import React, { useState, useEffect, useRef } from 'react';
import type { FC, ReactElement } from 'react';

import soundFile from '../assets/soundFile.wav';

import type * as Types from './Sound.types';

const Sound: FC<Readonly<Types.SoundProps>> = ({ shouldTriggerSound }): ReactElement => {
  const [soundIsPlaying, setSoundIsPlaying] = useState(false);
  const prevPlaySound = useRef(false);

  useEffect(() => {
    // do not trigger new sound start while previous sound is playing
    if (shouldTriggerSound && shouldTriggerSound !== prevPlaySound.current) {
      setSoundIsPlaying(true);

      window.setTimeout(() => {
        setSoundIsPlaying(false);
      }, 800);
    }

    prevPlaySound.current = shouldTriggerSound;
  }, [shouldTriggerSound]);

  if (!soundIsPlaying) {
    return null;
  }

  return (
    <audio
      autoPlay
      data-testid="audio-tag"
    >
      <source
        src={soundFile}
        type="audio/wav"
        data-testid="audio-file"
      />
    </audio>
  );
};

export default Sound;
