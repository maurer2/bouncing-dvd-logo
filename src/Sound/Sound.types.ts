export type SoundProps = {
  playSound: boolean;
}
