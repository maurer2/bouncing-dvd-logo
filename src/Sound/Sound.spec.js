import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import Sound from './Sound';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('Sound', () => {
  const wrapper = shallow(<Sound playSound={false} />);

  test('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ playSound: true });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
