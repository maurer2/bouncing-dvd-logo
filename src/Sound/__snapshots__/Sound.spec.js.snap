// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sound should match snapshot 1`] = `""`;
